system/Test/FilterTestTrait.php

```diff
@@ -0,0 +1,289 @@
+	 */
+	protected $response;
+
+	/**
+	 * The Filters configuration to use.
+	 * Extracted for access to aliases
+	 * during Filters::discoverFilters().
+	 *
+	 * @var FiltersConfig|null
+	 */
+	protected $filtersConfig;
+
+	/**
+	 * The prepared Filters library.
+	 *
+	 * @var Filters|null
+	 */
+	protected $filters;
+
+	/**
+	 * The default App and discovered
+	 * routes to check for filters.
+	 *
+	 * @var RouteCollection|null
+	 */
+	protected $collection;
+
+	//--------------------------------------------------------------------
+	// Staging
+	//--------------------------------------------------------------------
+
+	/**
+	 * Initializes dependencies once.
+	 *
+	 * @return void
+	 */
+	protected function setUpFilterTestTrait(): void
+	{
+		if ($this->doneFilterSetUp === false)
+		{
+			// Create our own Request and Response so we can
+			// use the same ones for Filters and FilterInterface
+			// yet isolate them from outside influence
+			$this->request  = $this->request ?? clone Services::request();
+			$this->response = $this->response ?? clone Services::response();
+
+			// Create our config and Filters instance to reuse for performance
+			$this->filtersConfig = $this->filtersConfig ?? config('Filters');
+			$this->filters       = $this->filters ?? new Filters($this->filtersConfig, $this->request, $this->response);
+
+			if (is_null($this->collection))
+			{
+				// Load the RouteCollection from Config to gather App route info
+				// (creates $routes using the Service as a starting point)
+				require APPPATH . 'Config/Routes.php';
+
+				$routes->getRoutes('*'); // Triggers discovery
+				$this->collection = $routes;
+			}
+
+			$this->doneFilterSetUp = true;
+		}
+	}
+
+	//--------------------------------------------------------------------
+	// Utility
+	//--------------------------------------------------------------------
+
+	/**
+	 * Returns a callable method for a filter position
+	 * using the local HTTP instances.
+	 *
+	 * @param FilterInterface|string $filter   The filter instance, class, or alias
+	 * @param string                 $position "before" or "after"
+	 *
+	 * @return callable
+	 */
+	protected function getFilterCaller($filter, string $position): callable
+	{
+		if (! in_array($position, ['before', 'after'], true))
+		{
+			throw new InvalidArgumentException('Invalid filter position passed: ' . $position);
+		}
+
+		if (is_string($filter))
+		{
+			// Check for an alias (no namespace)
+			if (strpos($filter, '\\') === false)
+			{
+				if (! isset($this->filtersConfig->aliases[$filter]))
+				{
+					throw new RuntimeException("No filter found with alias '{$filter}'");
+				}
+
+				$filter = $this->filtersConfig->aliases[$filter];
+			}
+
+			// Get an instance
+			$filter = new $filter();
+		}
+
+		if (! $filter instanceof FilterInterface)
+		{
+			throw FilterException::forIncorrectInterface(get_class($filter));
+		}
+
+		if ($position === 'before')
+		{
+			$request = clone $this->request;
```

Comment:
put this before the if so line 168 would be not needed

Reply:
Good suggestion, implementing.